lazy-pages/src/sys.rs

```diff
@@ -107,6 +113,174 @@ pub(crate) enum AccessedPagesInfo {
+        self.inner_access_provider.get_i64(name)
+    }
+
+    fn set_i64(&mut self, name: &str, value: i64) -> Result<(), GlobalsAccessError> {
+        self.inner_access_provider.set_i64(name, value)
+    }
+
+    fn as_any_mut(&mut self) -> &mut dyn Any {
+        unreachable!()
+    }
+}
+
+fn charge_gas_internal(
+    mut globals_access_provider: impl GlobalsAccesser,
+    global_gas: &str,
+    global_allowance: &str,
+    amount: u64,
+) -> Result<Status, Error> {
+    let mut sub_global = |name, value| {
+        let current_value = globals_access_provider.get_i64(name).ok()? as u64;
+        let (new_value, exceed) = current_value
+            .checked_sub(value)
+            .map(|val| (val, false))
+            .unwrap_or((0, true));
+        globals_access_provider
+            .set_i64(name, new_value as i64)
+            .ok()?;
+
+        log::trace!("Change global {name}: {current_value} -> {new_value}, exceeded: {exceed}");
+
+        Some(exceed)
+    };
+    if sub_global(global_gas, amount).ok_or(Error::CannotChargeGas)? {
+        return Ok(Status::GasLimitExceeded);
+    }
+    if sub_global(global_allowance, amount).ok_or(Error::CannotChargeGasAllowance)? {
+        return Ok(Status::GasAllowanceExceeded);
+    }
+    Ok(Status::Normal)
+}
+
+unsafe fn charge_gas(globals_ctx: GlobalsCtx, amount: u64) -> Result<Status, Error> {
+    match globals_ctx.globals_access_mod {
+        GlobalsAccessMod::WasmRuntime => {
+            let instance = (globals_ctx.globals_access_ptr as *mut SandboxInstance)
+                .as_mut()
+                .ok_or(Error::HostInstancePointerIsInvalid)?;
+            charge_gas_internal(
+                GlobalsAccessWasmRuntime { instance },
+                &globals_ctx.global_gas_name,
+                &globals_ctx.global_allowance_name,
+                amount,
+            )
+        }
+        GlobalsAccessMod::NativeRuntime => {
+            let inner_access_provider = (globals_ctx.globals_access_ptr
+                as *mut &mut dyn GlobalsAccesser)
+                .as_mut()
+                .ok_or(Error::DynGlobalsAccessPointerIsInvalid)?;
+            charge_gas_internal(
+                GlobalsAccessNativeRuntime {
+                    inner_access_provider,
+                },
+                &globals_ctx.global_gas_name,
+                &globals_ctx.global_allowance_name,
+                amount,
+            )
+        }
+    }
+}
+
+unsafe fn charge_for_pages(
+    ctx: &mut RefMut<LazyPagesExecutionContext>,
+    pages: PagesIterInclusive<LazyPage>,
+    is_write: bool,
+) -> Result<Status, Error> {
+    let globals_ctx = if let Some(ctx) = ctx.globals_ctx.as_ref() {
+        ctx.clone()
+    } else {
+        return Ok(Status::Normal);
+    };
+
+    let mut amount = 0u64;
+    for page in pages
+        .map(|page| page.to_page::<GranularityPage>())
+        .collect::<BTreeSet<_>>()
+        .into_iter()
+    {
+        let amount_for_page = if is_write {
```

Comment:
over complicated if
asking to move into separate fns with sections that separated using return statement

Reply:
Done